tests/conformance/pubsub/pubsub.go

```diff
@@ -0,0 +1,127 @@
+package pubsub
+
+import (
+	"sort"
+	"strconv"
+	"strings"
+	"testing"
+	"time"
+
+	"github.com/dapr/components-contrib/pubsub"
+	"github.com/dapr/components-contrib/tests/conformance/utils"
+	"github.com/stretchr/testify/assert"
+)
+
+const (
+	defaultPubsubName      = "pubusub"
+	defaultTopicName       = "testTopic"
+	defaultMessageCount    = 10
+	defaultMaxReadDuration = 10 * time.Millisecond
+)
+
+type TestConfig struct {
+	allOperations     bool
+	operations        []string
+	componentName     string
+	pubsubName        string
+	testTopicName     string
+	publishMetadata   map[string]string
+	subscribeMetadata map[string]string
+	messageCount      int
+	maxReadDuration   time.Duration
+}
+
+func NewTestConfig(componentName string, allOperations bool, operations []string, config map[string]string) TestConfig {
+	tc := TestConfig{
+		componentName:     componentName,
+		allOperations:     allOperations,
+		operations:        operations,
+		pubsubName:        defaultPubsubName,
+		testTopicName:     defaultTopicName,
+		messageCount:      defaultMessageCount,
+		maxReadDuration:   defaultMaxReadDuration,
+		publishMetadata:   map[string]string{},
+		subscribeMetadata: map[string]string{},
+	}
+	sort.Strings(tc.operations)
+	for k, v := range config {
+		if k == "pubsubName" {
+			tc.pubsubName = v
+		}
+		if k == "testTopicName" {
+			tc.testTopicName = v
+		}
+		if k == "messageCount" {
+			val, err := strconv.Atoi(v)
+			if err == nil {
+				tc.messageCount = val
+			}
+		}
+		if k == "maxReadDuration" {
+			val, err := strconv.Atoi(v)
+			if err == nil {
+				tc.maxReadDuration = time.Duration(val) * time.Millisecond
+			}
+		}
+		if strings.HasPrefix(k, "publish_") {
+			tc.publishMetadata[strings.Replace(k, "publish_", "", 1)] = v
+		}
+		if strings.HasPrefix(k, "subscribe_") {
+			tc.subscribeMetadata[strings.Replace(k, "subscribe_", "", 1)] = v
+		}
+	}
+
+	return tc
+}
+
+func ConformanceTests(t *testing.T, props map[string]string, pubusub pubsub.PubSub, config TestConfig) {
+	// Properly close connection to pubsub
+	defer pubusub.Close()
+
+	actualReadCount := 0
+	if utils.HasOperation(config.allOperations, config.operations, "init") {
```

Comment:
Can this be a method on `TestConfig` instead of `utils`?

Reply:
then it would need to be repeated for each TestConfig.

Reply:
I guess that is slightly better than repeating for each `if`, right? you can also use the utils in the `TestConfig.HasOperation(...)` if you don't like embedding the checks.